lucille-core/src/main/java/com/kmwllc/lucille/stage/JSoupExtractor.java

```diff
@@ -0,0 +1,103 @@
+ * </p>
+ */
+public class JSoupExtractor extends Stage {
+
+  private String filePathField;
+  private Map<String, Object> destinationFields;
+  private String byteArrayField;
+
+  public JSoupExtractor(Config config) {
+    super(config,
+        new StageSpec().withOptionalProperties("filePathField", "byteArrayField").withRequiredParents("destinationFields"));
+
+    this.destinationFields = config.getConfig("destinationFields").root().unwrapped();
+    this.filePathField = ConfigUtils.getOrDefault(config, "filePathField", null);
+    this.byteArrayField = ConfigUtils.getOrDefault(config, "byteArrayField", null);
+  }
+
+  @Override
+  public void start() throws StageException {
+    if (this.filePathField != null && this.byteArrayField != null) {
+      throw new StageException("Stage cannot have both filePathField and byteArrayField specified");
+    }
+
+    for (Map.Entry<String, Object> entry : destinationFields.entrySet()) {
+      Object value = entry.getValue();
+      if (value instanceof Map) {
+        Map<String, String> map = (Map<String, String>) value;
+        if (map.keySet().size() != 2)
+          throw new StageException("map must contain two fields");
+        if (!map.containsKey("attribute") || !map.containsKey("selector"))
+          throw new StageException("map must contain a attribute and selector fields");
+      } else if (!(value instanceof String)) {
+        throw new StageException("destinationFields mapping can only have strings or a map as its values");
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    String field = filePathField != null ? filePathField : byteArrayField;
+
+    if (doc.has(field)) {
+      try {
+        org.jsoup.nodes.Document jsoupDoc = filePathField != null ? Jsoup.parse(new File(doc.getString(filePathField)))
+            : Jsoup.parse(new String(doc.getBytes(byteArrayField), StandardCharsets.UTF_8));
+        System.out.println(super.getName() + " " + destinationFields.entrySet());
+        for (Map.Entry<String, Object> entry : destinationFields.entrySet()) {
+          Object value = entry.getValue();
+
+          if (value instanceof String) {
+            // if the user specified a `desination -> selector` mapping
+            doc.update(entry.getKey(), UpdateMode.OVERWRITE, jsoupDoc.select((String) value).eachText().toArray(new String[0]));
+          } else {
+            // if the user specified a `destination -> {selector: selector, attribute: attribute}` mapping
+            Map<String, String> config = (Map<String, String>) value;
+            doc.update(entry.getKey(), UpdateMode.OVERWRITE, jsoupDoc.select(config.get("selector")).eachAttr(config.get("attribute")).toArray(new String[0]));
+          }
+
+        }
+      } catch (IOException e) {
+        throw new StageException("File parse failed: " + e.getMessage());
```

Comment:
about the exception message -- we weren't necessarily parsing a file... for example, we could be parsing a byte[] that was set by a FetchUri stage

so... message should say something like "HTML parsing failed"

Reply:
The  IOException is only thrown when the file parsing fails. Jsoup.parse for a String does not throw any declared exceptions

Reply:
ok, I'd suggest to put the try/catch only around the smaller block of code where an exception can be thrown -- only the line where we're calling Jsoup.parse. Also add a brief comment saying the IOException is only thrown when parsing a file